File: test-automation-core/src/main/java/com/nortal/test/core/services/report/cucumber/SkippedScenarios.java
package com.nortal.test.core.services.report.cucumber;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import io.cucumber.testng.Pickle;
import io.cucumber.testng.PickleWrapper;

public class SkippedScenarios {
    private final Map<String, List<PickleWrapper>> skippedScenariosByReportableName;
    private final Set<String> featureNames;

    public SkippedScenarios(final Map<String, List<PickleWrapper>> skippedScenariosByFeatureName) {
        skippedScenariosByReportableName = new HashMap<>();
        if(skippedScenariosByFeatureName == null) {
            featureNames = Collections.emptySet();
            return;
        }
        List<PickleWrapper> pickles = skippedScenariosByFeatureName.values().stream()
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
        final HashMap<String, List<PickleWrapper>> skippedScenariosByTag = createSkippedScenariosByTag(pickles);

        skippedScenariosByReportableName.putAll(skippedScenariosByFeatureName);
        featureNames = skippedScenariosByFeatureName.keySet();
        skippedScenariosByReportableName.putAll(skippedScenariosByTag);
    }

    private HashMap<String, List<PickleWrapper>> createSkippedScenariosByTag(final List<PickleWrapper> pickles) {
        final HashMap<String, List<PickleWrapper>> skippedScenariosByTag = new HashMap<>();
        pickles.forEach(pickleWrapper -> pickleWrapper.getPickle().getTags().forEach( tag ->
                        skippedScenariosByTag.compute(tag, (key, pickleWrappers) -> {
                            List<PickleWrapper> skippedPickles = new ArrayList<>();
                            if(pickleWrappers != null) {
                                skippedPickles = pickleWrappers;
                            }
                            skippedPickles.add(pickleWrapper);
                            return skippedPickles;
                        })
                                                                                    ));
        return skippedScenariosByTag;
    }

    public Set<String> getFeatureNames() {
        return featureNames;
    }

    public List<String> getScenarioNamesByReportableName(final String name) {
        return skippedScenariosByReportableName.getOrDefault(name, Collections.emptyList()).stream()
                .map(pickleWrapper -> pickleWrapper.getPickle().getName())
                .collect(Collectors.toList());
    }

    public List<Pickle> getScenariosByReportableName(final String name) {
        return skippedScenariosByReportableName.getOrDefault(name, Collections.emptyList()).stream()
                .map(PickleWrapper::getPickle)
                .collect(Collectors.toList());
    }

    public int getTotalScenarioCount() {
        // Casting to int because we should never have more than 2147483647 scenarios.
        // The tests would never finish.
        return (int) skippedScenariosByReportableName.values().stream()
                .flatMap(Collection::stream)
                .distinct()
                .count();
    }

}

File: test-automation-core-old/src/main/java/com/nortal/test/core/services/report/cucumber/SkippedScenarios.kt
package com.nortal.test.core.services.report.cucumber

import io.cucumber.testng.Pickle
import io.cucumber.testng.PickleWrapper
import java.util.ArrayList
import java.util.HashMap
import java.util.function.BiFunction
import java.util.function.Consumer
import java.util.stream.Collectors

class SkippedScenarios(skippedScenariosByFeatureName: Map<String, MutableList<PickleWrapper>>?) {
    private val skippedScenariosByReportableName: MutableMap<String, MutableList<PickleWrapper>>
    var featureNames: Set<String>

    init {
        skippedScenariosByReportableName = HashMap()
        if (skippedScenariosByFeatureName == null) {
            featureNames = emptySet()
        } else {
            val pickles = skippedScenariosByFeatureName!!.values.stream()
                .flatMap { obj: List<PickleWrapper> -> obj.stream() }
                .collect(Collectors.toList())
            val skippedScenariosByTag = createSkippedScenariosByTag(pickles)
            skippedScenariosByReportableName.putAll(skippedScenariosByFeatureName)
            featureNames = skippedScenariosByFeatureName.keys
            skippedScenariosByReportableName.putAll(skippedScenariosByTag)
        }
    }

    private fun createSkippedScenariosByTag(pickles: List<PickleWrapper>): HashMap<String, MutableList<PickleWrapper>> {
        val skippedScenariosByTag = HashMap<String, MutableList<PickleWrapper>>()
        pickles.forEach(Consumer { pickleWrapper: PickleWrapper ->
            pickleWrapper.pickle.tags.forEach(
                Consumer { tag: String ->
                    skippedScenariosByTag.compute(tag, BiFunction { key: String?, pickleWrappers: MutableList<PickleWrapper>? ->
                        var skippedPickles: MutableList<PickleWrapper> = ArrayList()
                        if (pickleWrappers != null) {
                            skippedPickles = pickleWrappers
                        }
                        skippedPickles.add(pickleWrapper)
                        skippedPickles
                    })
                }
            )
        })
        return skippedScenariosByTag
    }

    fun getScenarioNamesByReportableName(name: String?): List<String> {
        return skippedScenariosByReportableName.getOrDefault(name, emptyList()).stream()
            .map { pickleWrapper: PickleWrapper -> pickleWrapper.pickle.name }
            .collect(Collectors.toList())
    }

    fun getScenariosByReportableName(name: String?): List<Pickle> {
        return skippedScenariosByReportableName.getOrDefault(name, emptyList()).stream()
            .map { obj: PickleWrapper -> obj.pickle }
            .collect(Collectors.toList())
    }

    // Casting to int because we should never have more than 2147483647 scenarios.
    // The tests would never finish.
    val totalScenarioCount: Int
        get() =// Casting to int because we should never have more than 2147483647 scenarios.
            // The tests would never finish.
            skippedScenariosByReportableName.values.stream()
                .flatMap { obj: List<PickleWrapper> -> obj.stream() }
                .distinct()
                .count().toInt()
}